[EN] add eBPF (#2050)
cncf/glossary: content/en/ebpf.md

---
title: eBPF
status: Completed
category: architecture
---

## What it is

> 

Reviewer: ```suggestion
eBPF, or extended Berkeley Packet Filter, is a technology that allows applications to run within the kernel of a Linux system without changing the kernel source code or load Linux kernel modules.
```
I think the first sentence must clearly state what it is. We can then break that down and explain.

---
title: eBPF
status: Completed
category: architecture
---

## What it is

> Linux system has two spaces: kernel and user space. Kernel space is the core of the operating system and it is the only part that has unlimited access to the hardware. 

Reviewer: ```suggestion
A Linux system has two spaces: the kernel and the user space. 
The kernel represents the operating system's core and is the only part with unlimited access to the hardware. 
```

> Usually, applications reside in the user space and if they need some higher permissions, they send a request to the kernel space for that.

Reviewer: ```suggestion
Usually, applications reside in the user space, and if they need higher permissions, they send a request to the kernel. 
```
Usually or always? And what does "higher permissions" mean? Can we elaborate?

Author: Removed "Usually" and explained the higher permissions.

> For applications that require more flexibility (exp. to access hardware), there is an approach called “Linux kernel modules”. It allows extending the kernel base without adding directly to the kernel source code. 

Reviewer: ```suggestion
For applications that require more flexibility, e.g., access hardware, the kernel can be extended with eBPF via a so-called "Linux kernel modules" approach.   
```

Reviewer: Are there other ways of taking a "Linus kernel module" approach that isn't eBPF?

> However, it also introduces some security risks.

Reviewer: ```suggestion
However, such an approach also introduces some security risks.
```

Reviewer: Risk applies to other approaches (non-eBFP) or all these approaches?

Author: I added some edit suggestions about this.

> eBPF stands for "extended Berkeley Packet Filter". 

Reviewer: ```suggestion
```
Moved this up

> It is a revolved technology that allows applications to run within the kernel space of a Linux system without the need to change the kernel source or load Linux kernel modules.

Reviewer: ```suggestion
```
Moved this up

However, since kernel modules operate within the kernel space, this approach introduces some security risks.

## Problem it addresses
Normally, applications run in user space, and if the application requires some privileges from the kernel (exp. to access some hardware), then it needs to send a call referred to as a “system call” and ask for that request from the kernel. 
This might be good enough for many scenarios, however, there are cases, in which developers require more flexibility for hardware accessibility. Observability, security, and networking features are some of the ideal examples to be implemented within the kernel space. 
One of the possibilities to reach such a goal is to use Linux kernel modules, which allow extending the kernel base without adding directly to the kernel source code. 
While using Linux kernel modules bring the mentioned benefits, it adds to the security risks because they operate within the kernel space and can make the kernel crash. 

> Kernel modules have elevated privileges and direct access to system resources, making them a potential target for attackers.

Reviewer 2: True, but may be a bit misleading. I'm not sure the kernel modules themselves are typically the targets of attacks, but rather improperly implemented or insecure kernel modules can be exploited to gain unauthorized access.

```suggestion
Kernel modules have elevated privileges and direct access to system resources, which, if not properly secured, can be exploited by attackers.
```

## How it helps
Compared to Linux Kernel modules, eBPF provides a more controlled and contained environment for executing user-defined programs. 
eBPF programs run in a sandboxed environment within the kernel, providing isolation and mitigating some risks. 
If a vulnerability or flaw is exploited in an eBPF program, its impact is generally limited to the sandboxed environment.
Moreover, before an eBPF program can start running in the kernel, it has to pass some verifications. 

> The verifier component checks all the potential possibilities that would happen as a result of running the eBPF program. 

Reviewer 2: We could be a bit more specific here, the eBPF verifier does more than checking for potential outcomes. This suggestion may come at the cost of understanding for more non-technical folks though (adding a couple terms).
```suggestion
The verifier component checks the eBPF program for any potential safety violations, such as out-of-bounds memory access, infinite loops, and the usage of unauthorized kernel functions.
```

> This way, it makes sure that the program would not go into an infinitive loop and would not cause a kernel crash. 

Reviewer 2: ```suggestion
This way, it ensures that the program will not enter an infinite loop and will not cause a kernel crash.
```

---
title: eBPF
status: Completed
category: architecture
---

## What it is

eBPF, or extended Berkeley Packet Filter, is a technology that allows small, sandboxed programs or scripts to be run in the kernel space of a Linux system without having to change the kernel's source code or load Linux kernel modules.

A Linux system has two spaces: the kernel and the user space. 
The kernel represents the operating system's core and is the only part with unlimited access to the hardware. 

> 

Reviewer 3: From my personal taste, we go too much into detail in the following sentences, especially as we talk about some of those points in the next chapter. What do you think? Maybe we can shrink the next sentences to just add something like: 
Applications reside in the user space but sometimes require deeper access to the operating system than the default kernel provides. Linux kernel modules can extend the kernel's default functionality allowing applications to get deeper access to underlying components. Unfortunately, this could introduce security risks as well.

Author: Thanks, @iamNoah1 for your suggestions.

I understand your point here and I agree that it has become long and what you suggested summarises it. However, the first time that I encountered the term, it was pretty hard for me personally to understand it. That's why I tried to break it down into smaller and simpler definitions. What do you think if we mix the two, making it short, but still explaining a bit more? Something like:

> Applications reside in the user space, and when they need higher permissions, they send a request to the kernel. 
For applications that require more flexibility, such as direct hardware access, the kernel can be extended via what is known as the "Linux kernel modules" approach. It can extend the kernel's default functionality allowing applications deeper access to the underlying components. Unfortunately, this could introduce security risks as well.

Reviewer 3: @MaryamTavakkoli Yes that would be fine from my side :)

Reviewer 3: I am confused, did you apply the suggestions?!

Observability, security, and networking features are some of the ideal examples to be implemented within the kernel space.
One of the possibilities to reach such a goal is to use Linux kernel modules, which allow extending the kernel base without adding directly to the kernel source code. 
While using Linux kernel modules bring the mentioned benefits, it adds to the security risks because they operate within the kernel space and can make the kernel crash. 
Kernel modules have elevated privileges and direct access to system resources, which, if not properly secured, can be exploited by attackers.

## How it helps
Compared to Linux Kernel modules, eBPF provides a more controlled and contained environment for executing user-defined programs. 
eBPF programs run in a sandboxed environment within the kernel, providing isolation and mitigating some risks. 
If a vulnerability or flaw is exploited in an eBPF program, its impact is generally limited to the sandboxed environment.
Moreover, before an eBPF program can start running in the kernel, it has to pass some verifications. 

> The verifier component checks the eBPF program for any potential safety violations, such as out-of-bounds memory access, infinite loops, and the usage of unauthorized kernel functions.

Reviewer 3: ```suggestion
The verifier component checks the eBPF program for any potential safety violations, 
such as out-of-bounds memory access, infinite loops, and the usage of unauthorized kernel functions.
```

---
title: eBPF
status: Completed
category: architecture
---

## What it is

> eBPF, or extended Berkeley Packet Filter, is a technology that allows small, sandboxed programs or scripts to be run in the kernel space of a Linux system without having to change the kernel's source code or load Linux kernel modules.

Reviewer: ```suggestion
eBPF, or extended Berkeley Packet Filter, is a technology that allows small, sandboxed programs or scripts to run in the kernel space of a Linux system without having to change the kernel's source code or load Linux kernel modules.
```

A Linux system has two spaces: the kernel and the user space. 
The kernel represents the operating system's core and is the only part 
with unlimited access to the hardware. 

Applications reside in the user space, and when they need higher permissions, 
they send a request to the kernel.
For applications that require more flexibility, such as direct hardware 
access, the kernel can be extended via what is known as the "Linux 

> kernel modules" approach. It can extend the kernel's default functionality

Reviewer: ```suggestion
kernel modules" approach. This approach extends the kernel's default functionality
```

> kernel modules" approach. This approach extends the kernel's default functionality

Reviewer: Question: reading through "the problem it addresses" it seems like eBFP is not a "Linux kernel model", but an alternative...or am I getting this wrong? If so, we could probably say below: "However, this approach also introduces security risks making eBFP an attractive alternative."

Reviewer 3: yes.

Author: Applied this as well.

 allowing applications deeper access to the underlying components. 
 Unfortunately, this could introduce security risks as well.

## Problem it addresses

> Normally, applications run in user space, and if the application requires some privileges from the kernel (exp. to access some hardware), 

Reviewer: ```suggestion
Typically, applications run in the user space, and if the application requires some privileges from the kernel (e.g., to access some hardware), 
```

> Typically, applications run in the user space, and if the application requires some privileges from the kernel (e.g., to access some hardware), 

Reviewer 3: code can also run in the kernel space, but I think this is too much detail. When we talk about applications we can assume that we talk about code that runs in the user space.

> then it needs to send a call referred to as a “system call” and ask for that request from the kernel. 

Reviewer: ```suggestion
it requests it from the kernel via a so-called "system call."  
```

> This might be sufficient for many scenarios, however, there are cases where developers require more flexibility for low-level system access.

Reviewer: ```suggestion
In most cases, this approach works just fine. However, there are instances where developers require more flexibility for low-level system access.
```

Observability, security, and networking features are some of the ideal examples to be implemented within the kernel space.

> One of the possibilities to reach such a goal is to use Linux kernel modules, which allow extending the kernel base without adding directly to the kernel source code. 

Reviewer: Making that sentence a little shorter but 100% sure what it means. Does it mean without modifying the kernel source code?
```suggestion
To achieve that, we can use Linux kernel modules, extending the kernel base without adding directly to the kernel source code. 
```

Author: As I understand it, there is specific code for the kernel module but it is not part of the main kernel source code. The kernel module could be loaded at runtime if needed and extend the functionality of the kernel.

Reviewer 3: yes, this is correct. The suggestion does not change the statement, so should be fine.

> While using Linux kernel modules bring the mentioned benefits, it adds to the security risks because they operate within the kernel space and can make the kernel crash. 

Reviewer: I assume this would lead to the entire machine crashing (?) In any case, we should explain why crashing the kernel is _really_ bad.
```suggestion
While there are benefits to using Linux kernel modules, it also introduces security risks. 
Because they operate within the kernel space, Linux kernel modules can crash the kernel, and when the kernel crashes, so does the entire machine.
```

Reviewer 3: I think this is correct.

> Kernel modules have elevated privileges and direct access to system resources, which, if not properly secured, can be exploited by attackers.

Reviewer: Did I understand this correctly? There are two risks: 1) the risk of crashing the kernel, and 2) the risk of giving attackers direct access to system resources 
```suggestion
Additionally, kernel modules have elevated privileges and direct access to system resources. And if not properly secured, attackers can exploit these.
```

Reviewer 3: Also this seems to be correct.

---
title: eBPF
status: Completed
category: architecture
---

## What it is

eBPF, or extended Berkeley Packet Filter, is a technology that allows small, sandboxed programs or scripts to run in the kernel space of a Linux system without having to change the kernel's source code or load Linux kernel modules.

A Linux system has two spaces: the kernel and the user space. 
The kernel represents the operating system's core and is the only part 
with unlimited access to the hardware. 

Applications reside in the user space, and when they need higher permissions, 
they send a request to the kernel.
For applications that require more flexibility, such as direct hardware 
access, the kernel can be extended via what is known as the "Linux 
kernel modules" approach. This approach extends the kernel's default functionality,
 allowing applications deeper access to the underlying components. 

>  However, this approach also introduces security risks, making eBFP an attractive alternative.

Reviewer 3: ```suggestion
 However, this approach also introduces security risks, making eBPF an attractive alternative.
```

## Problem it addresses
Typically, applications run in the user space, and if the application requires some privileges from the kernel (e.g., to access some hardware), 
it requests it from the kernel via a so-called "system call."  
In most cases, this approach works just fine. However, there are instances where developers require more flexibility for low-level system access.
Observability, security, and networking features are good examples.

> To achieve that, we can use Linux kernel modules, extending the kernel base without adding directly to the kernel source code. 

Reviewer: I think this is what is meant. Otherwise, not sure what "adding to the kernel source code" means.
```suggestion
To achieve that, we can use Linux kernel modules, extending the kernel base without modifying the kernel source code. 
```

While there are benefits to using Linux kernel modules, it also introduces security risks. 
Because they operate within the kernel space, Linux kernel modules can crash the kernel, and when the kernel crashes, so does the entire machine.
Additionally, kernel modules have elevated privileges and direct access to system resources. And if not properly secured, attackers can exploit these.

## How it helps

> Compared to Linux Kernel modules, eBPF provides a more controlled and contained environment for executing user-defined programs. 

Reviewer: Not incorrect, but since this is an eBPF definition, I'd lead the sentence with eBPF.
```suggestion
eBPF provides a more controlled and contained environment for executing user-defined programs than Linux Kernel modules.
```

> eBPF programs run in a sandboxed environment within the kernel, providing isolation and mitigating some risks. 

Reviewer: ```suggestion
It runs in a sandboxed environment within the kernel, providing isolation and mitigating risk. 
```

If a vulnerability or flaw is exploited in an eBPF program, its impact is generally limited to the sandboxed environment.
Moreover, before an eBPF program can start running in the kernel, it has to pass some verifications. 

> The verifier component checks the eBPF program for any potential safety violations, 

Reviewer: ```suggestion
The verifier component checks the eBPF program for potential safety violations, 
```

> such as out-of-bounds memory access, infinite loops, and the usage of unauthorized kernel functions.

Reviewer: ```suggestion
such as out-of-bounds memory access, infinite loops, and unauthorized kernel functions.
```

> This way, it ensures that the program will not enter an infinite loop and will not cause a kernel crash.

Reviewer: ```suggestion
This way, it ensures the program won't enter an infinite loop and cause a kernel crash.
```

> These safety controls make eBPF a more secure option to run applications in the Linux kernel compared to the Linux kernel modules.

Reviewer: ```suggestion
These safety controls make eBPF a more secure option to run applications in the Linux kernel than the Linux kernel modules.
```
